### PR TITLE
FEATURE: add supported interval for okex

### PR DESCRIPTION
**Pull request by MengShue**
1. Extract from my old PR - just add supported interval.
2. add its unit test.

### CODE REVIEW

File: pkg/exchange/okex/query_kline_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+	now := time.Now()
+	// test supported interval - minute
+	klineDetail, err := e.QueryKLines(context.Background(), queryOrder.Symbol, types.Interval("1m"), types.KLineQueryOptions{
+		Limit:   50,
+		EndTime: &now})
+	if assert.NoError(t, err) {
+		assert.NotEmpty(t, klineDetail)
+	}
+	t.Logf("kline detail: %+v", klineDetail)
```
Comment (bailantaotao) [inline pkg/exchange/okex/query_kline_test.go:33]: rm, the log will increase the storage size

File: pkg/exchange/okex/query_kline_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	e := New(key, secret, passphrase)
+
+	queryOrder := types.OrderQuery{
+		Symbol: "BTC-USDT",
+	}
+
+	now := time.Now()
+	// test supported interval - minute
+	klineDetail, err := e.QueryKLines(context.Background(), queryOrder.Symbol, types.Interval("1m"), types.KLineQueryOptions{
```
Comment (bailantaotao) [inline pkg/exchange/okex/query_kline_test.go:27]: Why you pass customizing variable 1m?
Using types.Interval1M is preferable to customizing the interval.

File: pkg/exchange/okex/query_kline_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		Limit:   50,
+		EndTime: &now})
+	if assert.NoError(t, err) {
+		assert.NotEmpty(t, klineDetail)
```
Comment (bailantaotao) [inline pkg/exchange/okex/query_kline_test.go:31]: Can we compare what the klineDetail is expecting?

File: pkg/exchange/okex/convert.go [hunk 4]
```diff
@@ -206,11 +206,23 @@ func toGlobalOrderType(orderType okexapi.OrderType) (types.OrderType, error) {
...
+func toLocalInterval(interval types.Interval) (string, error) {
+	if _, ok := SupportedIntervals[interval]; !ok {
+		return "", fmt.Errorf("interval %s is not supported", interval)
+	}
+
+	switch i := interval.String(); {
+	case strings.HasSuffix(i, "m"):
+		return i, nil
+	case strings.HasSuffix(i, "mo"):
+		return "1M", nil
```
[the remark this answers is not in the document]
Comment (MengShue) [reply pkg/exchange/okex/convert.go:218]: Hong Kong time, ref: https://www.okx.com/docs-v5/en/#order-book-trading-market-data-get-candlesticks

File: pkg/exchange/okex/query_kline_test.go (hunk 1 above)
Comment (MengShue) [reply pkg/exchange/okex/query_kline_test.go:33]: ok, but I think we didn't do any unit test while deploying ?

Comment (bailantaotao) [reply pkg/exchange/okex/query_kline_test.go:33]: The unit test is necessary.
The log messages are not necessary in unit test. Unless you want to debug.

Comment (bailantaotao) [reply pkg/exchange/okex/query_kline_test.go]: Oh, I understand why you are said ` but I think we didn't do any unit test while deploying `. If you retrieve data from the actual server, it is considered an integration test, not a unit test.

A unit test typically involves creating a request and invoking the mock server, then confirming that the response matches the expected outcome.

File: pkg/exchange/okex/convert.go [hunk 5]
```diff
@@ -206,11 +206,36 @@ func toGlobalOrderType(orderType okexapi.OrderType) (types.OrderType, error) {
...
+	case strings.HasSuffix(i, "m"):
+		return i, nil
+	case strings.HasSuffix(i, "mo"):
+		return "1Mutc", nil
+	default:
+		hdwRegex := regexp.MustCompile("\\d+[dw]$")
+		if hdwRegex.Match([]byte(i)) {
+			return strings.ToUpper(i) + "utc", nil
+		}
+		hdwRegex = regexp.MustCompile("(\\d+)[h]$")
```
Comment (c9s) [inline pkg/exchange/okex/convert.go:224]: 這邊為何要換 pattern?

File: pkg/exchange/okex/convert.go [hunk 6]
```diff
@@ -206,11 +206,36 @@ func toGlobalOrderType(orderType okexapi.OrderType) (types.OrderType, error) {
...
-	var re = regexp.MustCompile(`\d+[hdw]`)
-	return re.ReplaceAllStringFunc(src, func(w string) string {
-		return strings.ToUpper(w)
-	})
+func toLocalInterval(interval types.Interval) (string, error) {
+	if _, ok := SupportedIntervals[interval]; !ok {
+		return "", fmt.Errorf("interval %s is not supported", interval)
+	}
+
+	switch i := interval.String(); {
```
Comment (c9s) [inline pkg/exchange/okex/convert.go:214]: 除了 index 之外盡量不要用 `i` 這個 variable

File: pkg/exchange/okex/query_kline_test.go (hunk 3 above)
Comment (c9s) [reply pkg/exchange/okex/query_kline_test.go:31]: I think the EndTime varies, so the kline data will be different

File: pkg/exchange/okex/convert.go (hunk 5 above)
Comment (MengShue) [reply pkg/exchange/okex/convert.go:224]: [1m/3m/5m/15m/30m/1H/2H/4H] + [/6Hutc/12Hutc/1Dutc/2Dutc/3Dutc/1Wutc/1Mutc/3Mutc]
4H 以下不用加utc, 從6H 以上要加utc, 所以 H 單獨拿出來做一新的 REGEX, 並搭配 group 抓digits 使用，`>= 6` 加上utc, `< 6` 則不用

File: pkg/exchange/okex/convert.go (hunk 6 above)
Comment (MengShue) [reply pkg/exchange/okex/convert.go:214]: 好

File: pkg/exchange/okex/types.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+		"1H":     types.Interval1h,
+		"2H":     types.Interval2h,
+		"4H":     types.Interval4h,
+		"6Hutc":  types.Interval6h,
+		"12Hutc": types.Interval12h,
+		"1Dutc":  types.Interval1d,
+		"3Dutc":  types.Interval3d,
+		"1Wutc":  types.Interval1w,
+		"1Mutc":  types.Interval1mo,
+	}
```
[the remark this answers is not in the document]
Comment (c9s) [reply pkg/exchange/okex/types.go:39]: 就查表就好

[the remark this answers is not in the document]
Comment (MengShue) [reply pkg/exchange/okex/types.go:39]: Updated

**Review by c9s — APPROVED** [review]

File: pkg/exchange/okex/convert.go [hunk 8]
```diff
@@ -206,11 +205,16 @@ func toGlobalOrderType(orderType okexapi.OrderType) (types.OrderType, error) {
...
-	var re = regexp.MustCompile(`\d+[hdw]`)
-	return re.ReplaceAllStringFunc(src, func(w string) string {
-		return strings.ToUpper(w)
-	})
+func toLocalInterval(interval types.Interval) (string, error) {
+	if _, ok := SupportedIntervals[interval]; !ok {
+		return "", fmt.Errorf("interval %s is not supported", interval)
+	}
+
+	if in, ok := ToLocalInterval[interval]; ok {
```
Comment (bailantaotao) [inline pkg/exchange/okex/convert.go:213]: return error first, then return result, so you can 
```
in, ok := ToLocalInterval[interval]
if !ok {
    return "", fmt.Errorf("interval %s is not supported", interval)
}

return in, nil

```

Comment (bailantaotao) [reply pkg/exchange/okex/convert.go:213]: Can you add `local interval xxx` in errror msg?

**Review by bailantaotao — APPROVED** [review]
others good

**Pull request merged**